lib/compilers/r8.ts

```diff
@@ -0,0 +1,165 @@
+            const javaOptions = _.compact(
+                javaCompiler.prepareArguments(
+                    this.getClasspathArgument(),
+                    javaCompiler.getDefaultFilters(),
+                    {}, // backendOptions
+                    inputFilename,
+                    outputFilename,
+                    [], // libraries
+                    [], // overrides
+                ),
+            );
+            initialResult = await javaCompiler.runCompiler(
+                javaCompiler.getInfo().exe,
+                javaOptions,
+                this.filename(inputFilename),
+                javaCompiler.getDefaultExecOptions(),
+            );
+        } else if (this.lang.id === 'android-kotlin') {
+            const kotlinCompiler = unwrap(
+                global.handler_config.compileHandler.findCompiler('kotlin', this.kotlinId),
+            ) as KotlinCompiler;
+            outputFilename = kotlinCompiler.getOutputFilename(preliminaryCompilePath);
+            const kotlinOptions = _.compact(
+                kotlinCompiler.prepareArguments(
+                    this.getClasspathArgument(),
+                    kotlinCompiler.getDefaultFilters(),
+                    {}, // backendOptions
+                    inputFilename,
+                    outputFilename,
+                    [], // libraries
+                    [], // overrides
+                ),
+            );
+            initialResult = await kotlinCompiler.runCompiler(
+                kotlinCompiler.getInfo().exe,
+                kotlinOptions,
+                this.filename(inputFilename),
+                kotlinCompiler.getDefaultExecOptions(),
+            );
+        } else {
+            logger.error('Language is neither android-java nor android-kotlin.');
+        }
+
+        // R8 should not run if initial compile stage failed, the JavaCompiler
+        // result can be returned instead.
+        if (initialResult.code !== 0) {
+            return initialResult;
+        }
+
+        if (!execOptions) {
+            execOptions = this.getDefaultExecOptions();
+        }
+        if (!execOptions.customCwd) {
+            execOptions.customCwd = path.dirname(inputFilename);
+        }
+
+        // The items in 'options' before the source file are user inputs.
+        const sourceFileOptionIndex = options.findIndex(option => {
+            return option.endsWith('.java') || option.endsWith('.kt');
+        });
+
+        const files = await fs.readdir(preliminaryCompilePath);
+        const classFiles = files.filter(f => f.endsWith('.class'));
+        const r8Options = [
+            '-Dcom.android.tools.r8.enableKeepAnnotations=1',
+            '-cp',
+            this.compiler.exe, // R8 jar.
+            'com.android.tools.r8.R8',
+            ...this.getR8LibArguments(),
+            ...options.slice(0, sourceFileOptionIndex),
+            ...classFiles,
+        ];
```

Comment:
Adding r8 is cool! Thanks for doing this! However, doesn't r8 enable shrinking by default? If so, won't the user see nothing when they select the r8 complier? I can think of a few solutions:
1. Add @Keep to the Java/Kotlin code example
2. Add "-dontshrink" to the default options to disable shrinking for now. Later, we add an Android Java IDE that enables shrinking and supports a proguard file. https://developer.android.com/build/shrink-code#keep-code

Reply:
Yes, R8 enables shrinking by default, so there's a template provided for both Android Java and Android Kotlin (R8Example) that demonstrates how a keep annotation can be used.

I can add the keep annotation to the default example as well, what do you think?

Reply:
Oh, okay. I didn't know I can load another example other than the default. I think keeping the keep annotations in a separate example is fine. Maybe update the comment in default example to instruct users to load the r8 example when they want to use the r8 compiler?

Reply:
I've added a comment to the default example pointing to R8Example. Thanks!